Add types/mux.d.ts

dylanjha opened the pull request:
* Adds types 🎉 
* Removes erroneous requirement for `viewer_id` when querying list of view

Use this today with the `beta` tag:

* [`v3.0.0-beta.1`](https://www.npmjs.com/package/@mux/mux-node/v/3.0.0-beta.1)
* See Changelog notes: https://github.com/muxinc/mux-node-sdk/blob/dj/types/CHANGELOG.md#300

dylanjha commented on package.json:
```diff
@@ -14,29 +14,31 @@
   "main": "src/mux.js",
   "engines": {
-    "node": ">=8"
+    "node": ">=10"
```
This bumps the min node version to 10.

Node 8 has reached end-of-life: https://github.com/nodejs/Release#end-of-life-releases

<img width="967" alt="node-8-end-of-life_2020-05-29_14-44-23" src="https://user-images.githubusercontent.com/764988/83308291-5c6fab80-a1bb-11ea-8ef4-3ee626e8c8eb.png">


Unfortunately, this will require a major version bump but I think we should still do it. The reason this change is part of this PR is because the `eslint` version bump below (required to lint typescript stuff) requires node 10+

As @JacquiManzi remembered -- some [kook 2 years ago before they even worked at Mux needed Node 8 support](https://github.com/muxinc/mux-node-sdk/pull/4)... Ah how things come full circle 😄

dylanjha commented on .travis.yml:
```diff
@@ -1,9 +1,10 @@
   - '10'
+  - '12'
+  - '13'
+  - '14'
```
let's use travis to make sure our library works on newer node versions

coreymux commented on types/mux.d.ts:
```diff
@@ -0,0 +1,250 @@
+
+declare type Track = VideoTrack | AudioTrack | TextTrack;
+
+type MasterAcces = 'none' | 'temporary';
```
👀

dylanjha replied:
ty - I moved this type to it's own file and fixed the type

dylanjha commented:
Most folks who use this SDK are using the `Mux.Video` APIs, but I did notice one thing about the Data APIs that makes this SDK a bit difficult.

**All responses unwrap the `data` key.**

For video calls, this is great, instead of getting:

```js
{
data: {
    status: 'ready',
    id: 'KrkrIaMuQTPKQsrxgxaHK23NFHjsHbaO',
    duration: 10.083333,
    created_at: '1590102694',
    aspect_ratio: '16:9' 
    //
    // ...
    // 
  }
}
```

You get this, which is nicer:

```
{
    status: 'ready',
    id: 'KrkrIaMuQTPKQsrxgxaHK23NFHjsHbaO',
    duration: 10.083333,
    created_at: '1590102694',
    aspect_ratio: '16:9' 
    //
    // ...
    // 
  }
```

This is definitely more convenient.

However when using some of the data APIs this makes things more difficult, for example:

* All of the data endpoints include a `timeframe` key and a `total_row_count` key that get wiped out. For example [Breakdown](https://api-docs.mux.com/#breakdown), [Comparison](https://api-docs.mux.com/#comparison)
* This [Realtime `/metrics` endpoint](https://api-docs.mux.com/#real-time-get-3) includes data like this in a `meta` key, which seems like information you have to know:

```
 "meta": {
    "buckets": [
      {
        "start": 0,
        "end": 100
      },
      {
        "start": 100,
        "end": 500
      },
      {
        "start": 500,
        "end": 1000
      },
      {
        "start": 1000,
        "end": 2000
      },
    ],
    "bucket_unit": "milliseconds"
  },
```

----

I see in the docs we have a `'request'` and `'response'` event that gets emitted: https://github.com/muxinc/mux-node-sdk#request-and-response-events

This kind of addresses it, but I'm imagining if I write a script that is getting data from the realtime API and hitting a handful of routes asynchronously it wouldn't be the smoothest developer experience.

Since we have to make a major version upgrade in this branch (dropping Node 8 support), it might be worth revisiting this decision. One possible solution:

  * Don't pluck out the `data` key for routes within `/data/v1`? - Only do this for the video APIs. 

We don't have to address this here, but it's something I ran into that I wanted to get some thoughts on

JacquiManzi commented:
@dylanjha I think you're absolutely correct about our Data responses. We should definitely be returning responses that come back from our API in the same format. I think we should also do that for the Video responses. Thank you so much for bringing this up!

JacquiManzi commented on types/interfaces/AssetError.ts:
```diff
@@ -0,0 +1,4 @@
+export declare interface AssetError {
+  type: 'invalid_input' | string;
+  messages: string[];
```
I don't have a strong opinion either way on this, but should we be consistent with our use of `Array<string>` and `string[]`?

JacquiManzi commented on types/interfaces/CreateAssetParams.ts:
```diff
@@ -0,0 +1,14 @@
+
+export interface CreateAssetParams {
+  input: string | InputSettings[];
+  playback_policy?: PlaybackIdPolicy | (PlaybackIdPolicy)[];
```
There is probably a reason for the parenthesis here, right? `PlaybackIdPolicy)[]`?

dylanjha replied:
There's no reason, I think this syntactic style I copied over from some generated types based off our Open API spec -- good catch.

Ditto about the conflicting array syntax (`Array<string>` and `string[]`).

I'll make that consistent 👍

JacquiManzi commented on types/interfaces/CreateLiveStreamParams.ts:
```diff
@@ -0,0 +1,12 @@
+
+export interface CreateLiveStreamParams {
+  reconnect_window?: number;
+  playback_policy?: PlaybackIdPolicy | (PlaybackIdPolicy)[];
```
Same question as above for the parentheses here.

JacquiManzi commented on types/interfaces/VideoView.ts:
```diff
@@ -0,0 +1,114 @@
+  property_id?: number;
+  player_autoplay?: boolean;
+  player_height?: number;
+  asn?: number;
```
Just double checking that we return asn as a number.

JacquiManzi replied:
We do :)

JacquiManzi commented:
@dylanjha I went through this and everything seems good at a base level.

JacquiManzi commented on types/mux.d.ts:
```diff
@@ -0,0 +1,317 @@
+
+export declare class Assets extends Base {
+  create(params: CreateAssetParams): Promise<Asset>;
+  del(assetId: string): Promise<any>;
```
If we're going to a major version, we may want to remove the `remove` function in `base.js`:

```
  remove(...params) {
    // eslint-disable-next-line no-console
    console.warn(
      'The remove helper has been deprecated in favor of del. `remove` will no longer be available after the next major version bump (3.0).'
    );
    return this.del(...params);
  }
```

dylanjha replied:
good call

JacquiManzi commented on types/mux.d.ts:
```diff
@@ -0,0 +1,317 @@
+}
+
+export declare class LiveStreams extends Base {
+  create(params: CreateLiveStreamParams): Promise<any>;
```
Should this be `Promise<LiveStream>;` ?

dylanjha replied:
yep!

JacquiManzi commented on types/mux.d.ts:
```diff
@@ -0,0 +1,317 @@
+  RealTimeHistogramResponse,
+  RealTimeMetricsResponse,
+  RealTimeTimeseriesParams,
+  RealTimeTimeseriesResponse,
```
I _think_ we're missing `RealTimeDimensionsResponse`

JacquiManzi replied:
Nope it's there

dylanjha commented:
@JacquiManzi @mmcc this is up for review again!

- [ ] updated the implementation so that the `data` key only gets plucked off for `Video` (as we discussed)
- [ ] updated all interfaces for data responses accordingly
- [ ] updated integration tests to use the test account that @GeneticGenesis set up for [watch-this](https://github.com/muxinc/watch-this) which is DOPE!
- [ ] I enabled integration tests to run on travis, but sadly, we end up hitting Mux API rate limits 😢 ideas? (not a blocker)

### Release notes will look something like this for 3.0.0:

* added typescript support
* removed previously deprecated `remove` method ([use `del` instead](https://github.com/muxinc/mux-node-sdk/blob/master/CHANGELOG.md#soft-deprecations)) thanks @JacquiManzi 
* changed format of response for data endpoints
* remove erroneous requirement to pass in `viewer_id` when getting a list of views

Since we've had some folks asking for typescript support I think my plan is to release this on a `3.0.0-beta.0` tag and let people try this out before it's fully released. Thoughts? I've tested the typescript stuff in my own project but I'd like to make sure the developer experience is good "out in the wild"

mmcc commented on package.json:
```diff
@@ -21,8 +21,8 @@
     "docs": "esdoc",
     "test": "yarn test:unit",
```
I kinda think test should just be everything. We've got a bit of a weird split between unit/integration right now and especially if the default is mocked, I feel like the test options should be "run all the tests" and "run all the tests unmocked"

dylanjha replied:
yeah, good call 👍
